features/filesystem/bd/ChainingBlockDevice.cpp

```diff
@@ -0,0 +1,210 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ChainingBlockDevice.h"
+
+
+ChainingBlockDevice::ChainingBlockDevice(BlockDevice **bds, size_t bd_count)
+    : _bds(bds), _bd_count(bd_count)
+    , _read_size(0), _program_size(0), _erase_size(0), _size(0)
+{
+}
+
+static bool is_aligned(uint64_t x, uint64_t alignment)
+{
+    return (x / alignment) * alignment == x;
+}
+
+bd_error_t ChainingBlockDevice::init()
+{
+    _read_size = 0;
+    _program_size = 0;
+    _erase_size = 0;
+    _size = 0;
+
+    // Initialize children block devices, find all sizes and 
+    // assert that block sizes are similar. We can't do this in
+    // the constructor since some block devices may need to be
+    // initialized before they know their block size/count
+    for (size_t i = 0; i < _bd_count; i++) {
+        bd_error_t err = _bds[i]->init();
+        if (err) {
+            return err;
+        }
+
+        bd_size_t read = _bds[i]->read_size();
+        if (i == 0 || (read >= _read_size && is_aligned(read, _read_size))) {
+            _read_size = read;
+        } else if (_read_size > read && is_aligned(_read_size, read)) {
+            _read_size = _read_size;
```

Comment:
Just a style thing.

The purpose of this if statement is to choose the new value of `_read_size`. I can invert the second predicate as a condition for returning the error if you would like.